components/higher-order/with-safe-timeout/index.js

```diff
@@ -0,0 +1,63 @@
+/**
+ * WordPress dependencies
+ */
+import { Component } from '@wordpress/element';
+
+/**
+ * Browser dependencies
+ */
+const { clearTimeout, setTimeout } = window;
+
+/**
+ * A higher-order component used to provide and manage delayed function calls
+ * that ought to be bound to a component's lifecycle.
+ *
+ * @param {Component} OriginalComponent   Component requiring setTimeout
+ *
+ * @returns {Component}                   Wrapped component.
+ */
+function withSafeTimeout( OriginalComponent ) {
+	return class WrappedComponent extends Component {
+		constructor() {
+			super( ...arguments );
+			this.timeouts = [];
+			this.setTimeout = this.setTimeout.bind( this );
+			this.clearTimeout = this.clearTimeout.bind( this );
+			this.clear = this.clear.bind( this );
+		}
+
+		componentWillUnmount() {
+			this.timeouts.forEach( clearTimeout );
+		}
+
+		setTimeout( fn, delay ) {
+			const id = setTimeout( () => {
+				fn();
+				this.clear( id );
+			}, delay );
+			this.timeouts.push( id );
+			return id;
+		}
+
+		clearTimeout( id ) {
+			clearTimeout( id );
+			this.clear( id );
```

Comment:
Technically I think it'd be fine if the logic of `clear` was included here, than done separately, since `clearTimeout` on a finished timeout should be a noop (the reason I'm guessing you separated `clear`).

Reply:
Yep, the instinct was the separation, but we can merge.